[CodeComplete] Mark results as not recommended if global actor context doesn't match

ahoppen opened the pull request:
Annotate cross-actor references to global actors as `async` and mark them as not-recommended if they must be executed on a different actor than the current context is on.

Resolves rdar://75902598

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

rintaro commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1949,6 +1949,7 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(Evaluator &evaluator,
   if (auto CE = dyn_cast<ClosureExpr>(DC)) {
     if (CE->getBodyState() == ClosureExpr::BodyState::Parsed) {
       swift::typeCheckASTNodeAtLoc(CE->getParent(), CE->getLoc());
+      checkClosureActorIsolation(CE);
```
I'm concerned that `ActorIsolationChecker` walk into unrelated sibling closures or the children of this closure.
How about exposing `ActorIsolationChecker::determineClosureIsolation()` and call it directly?
```cpp
ActorIsolationChecker checker(closure->getParent());
CE->setActorIsolation(checker->determineClosureIsolation(CE));
```

rintaro replied:
Or, how about calling `checkFunctionActorIsolation()` (or `checkTopLevelActorIsolation()`) in `ide::typeCheckContextAt()` right after calling `typeCheckASTNodeAtLoc()`?

ahoppen replied:
Thanks. I didn’t realize that `TypeCheckASTNodeAtLocRequest` was called multiple times. 

I think it’ll be good to just set the current closure’s actor isolation as you proposed in the first comment because it minimizes the amount of actor isolation checking that needs to be done. I’ll update it.

rintaro commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4571,7 +4571,8 @@ struct CallbackCondition {
   CallbackCondition(const Decl *Subject, const CaseLabelItem *CaseItem) {
-    if (auto *EEP = dyn_cast<EnumElementPattern>(CaseItem->getPattern())) {
+    if (auto *EEP = dyn_cast<EnumElementPattern>(
+            CaseItem->getPattern()->getSemanticsProvidingPattern())) {
```
Unrelated changes?

ahoppen replied:
Oh, yes. I’ll remove it.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test macOS

ahoppen committed:
[CodeComplete] Mark results as not recommended if global actor context doesn't match

Annotate cross-actor references to global actors as `async` and mark them as not-recommended if they must be executed on a different actor than the current context is on.

Resolves rdar://75902598

ahoppen commented:
@swift-ci Please test

rintaro commented on test/IDE/complete_global_actorisolation.swift:
```diff
@@ -0,0 +1,168 @@
+    }
+  }
+}
+
```
Could you add a nested single expression closure test? e.g.
```swift
@MyGlobalActor func globalFuncOnMyGlobalActor() {}
@MyOtherGlobalActor func globalFuncOnMyOtherGlobalActor() {}

func foo<T>(_: () async -> T) {}

@MyGlobalActor func test() {
    foo {
        foo {
            <HERE>
        }
    }
}
```

ahoppen commented on test/IDE/complete_global_actorisolation.swift:
```diff
@@ -0,0 +1,168 @@
+class MyNonSendable {}
+struct MySendable {}
+
+actor MyActor {}
+
+@globalActor
+class MyGlobalActor {
+  static var shared = MyActor()
+}
```
I didn’t realize that. Thanks!

ahoppen commented:
@swift-ci Please test

rintaro commented on test/IDE/complete_global_actorisolation.swift:
```diff
@@ -0,0 +1,178 @@
+
+@globalActor
+actor MyGlobalActor {
+  static var shared = MyActor()
```
I believe this must be `MyGlobalActor` in the new syntax.

> The type of `shared` must be of the enclosing actor type

ahoppen replied:
Sorry, I missed that part. Fixed it now.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test macOS

ahoppen commented:
@swift-ci Please test macOS

ahoppen commented:
@swift-ci Please smoke test macOS

DougGregor commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -1949,6 +1949,19 @@ bool TypeCheckASTNodeAtLocRequest::evaluate(Evaluator &evaluator,
+      // TypeCheckASTNodeAtLocRequest is called multiple times, as described
+      // above.
+      auto ActorIsolation = determineClosureActorIsolation(CE);
+      CE->setActorIsolation(ActorIsolation);
```
Hmm, it feels a bit dicey setting the actor isolation for a closure in more than one place. Should we request'ify `determineClosureActorIsolation`?

ahoppen replied:
I agree that it’s not really pretty. But the problem is that the `ActorIsolationChecker` currently implicitly assumes that it’s walking the tree top to bottom. 

IMHO the proper solution would be to completely requesting `ActorIsolationChecker` so that it works bottom to top, requesting the actor isolation of an outer type as needed. But I felt that was too big of a change for Swift 5.5 and that’s why I stuck with the current, more local, approach.
